llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp

```diff
@@ -1926,6 +1926,43 @@ struct DSEState {
       if (InitC && InitC == StoredConstant)
         return MSSA.isLiveOnEntryDef(
             MSSA.getSkipSelfWalker()->getClobberingMemoryAccess(Def, BatchAA));
+
+      if (!Store)
+        return false;
+
+      // If there is a dominating condition, that ensures that the value
+      // being stored in a memory location is already present at the
+      // memory location, the store is a noop.
+      BasicBlock *StoreBB = DefI->getParent();
+      auto *StorePtr = Store->getOperand(1);
+
+      DomTreeNode *IDom = DT.getNode(StoreBB)->getIDom();
+      if (!IDom)
+        return false;
+
+      auto *TI = IDom->getBlock()->getTerminator();
+      ICmpInst::Predicate Pred;
+      BasicBlock *TrueBB, *FalseBB;
+
+      if (!match(TI, m_Br(m_ICmp(Pred, m_Load(m_Specific(StorePtr)),
+                                 m_Specific(StoredConstant)),
+                          TrueBB, FalseBB)))
+        return false;
+
+      MemoryAccess *LastMod =
+          MSSA.getSkipSelfWalker()->getClobberingMemoryAccess(Def, BatchAA);
+
+      DomTreeNode *CDom = DT.getNode(LastMod->getBlock());
```

Comment:
This check is too coarse-grained. It would allow something like

```
v = load p
clobber
if v == c {
   store x, p
}
```
You should fetch the MemoryAccess for the load and check whether the clobbering access dominates that one.

Reply:
Hi @nikic, was not able to understand the concern here. Can you please explain it in more detail?

I'm assuming the line inside the `if` block says `store v, p` (store the value `p` in `v`, if it does not then I do not understand the optimization that is supposed to be happening here. And the patch I pushed should not optimize this either since we would check for both `x` and `p` when we run a match on the icmp statement.)

This should not be optimized and will not be optimized with the current code. The current code checks for exactly StorePtr (v) and StoredConstant(p) in the `icmp` for the `br` that dominates the assignment, and whether these two are related by equality. If they are, and we have entered the current branch, and there were no clobbering accesses in between the current assignment and the conditional check, StorePtr should already contain the value that is being assigned to it.

And if any of these conditions are not met then we assume that the assignment is not redundant.

Reply:
Hi @nikic, while trying to test for this specific scenario I came accross the following observation:

If we have:
```
; *** IR Dump Before EarlyCSEPass on foo ***
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: none, inaccessiblemem: none) uwtable
define dso_local void @foo() local_unnamed_addr #0 {
  %1 = load i32, ptr @x, align 4, !tbaa !3
  store i32 7, ptr @x, align 4, !tbaa !3
  %2 = icmp eq i32 %1, 4
  %spec.store.select = select i1 %2, i32 4, i32 7
  store i32 %spec.store.select, ptr @x, align 4
  ret void
}
```

The EarlyCSEPass optimizes it to:
```
EarlyCSE DEAD STORE:   store i32 7, ptr @x, align 4, !tbaa !3  due to:   store i32 %spec.store.select, ptr @x, align 4
; *** IR Dump After EarlyCSEPass on foo ***
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: none, inaccessiblemem: none) uwtable
define dso_local void @foo() local_unnamed_addr #0 {
  %1 = load i32, ptr @x, align 4, !tbaa !3
  %2 = icmp eq i32 %1, 4
  %spec.store.select = select i1 %2, i32 4, i32 7
  store i32 %spec.store.select, ptr @x, align 4
  ret void
}
```

And calls the clobbering store after `load` a "dead store". Is this expected? 
If it is then the clobber that you mentioned would be optimized away as well wouldn't it? Please let me know if I'm misunderstanding something here.

Reply:
What is even more interesting is, even if we return the value stored at ptr x, we still optimize it in a similar manner (which I don't feel is the current behavior)

Say if we take:
```
@x = dso_local local_unnamed_addr global i32 0, align 4

; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: none, inaccessiblem
em: none) uwtable
define dso_local i32 @foo() local_unnamed_addr #0 {
  %1 = load i32, ptr @x, align 4, !tbaa !3
  store i32 7, ptr @x, align 4, !tbaa !3
  %2 = icmp eq i32 %1, 4
  br i1 %2, label %3, label %4

3:                                                ; preds = %0
  store i32 4, ptr @x, align 4, !tbaa !3
  br label %4

4:                                                ; preds = %3, %0
  %5 = load i32, ptr @x, align 4, !tbaa !3
  ret i32 %5
}
```

The code is optimized to:
```
@x = dso_local local_unnamed_addr global i32 0, align 4

; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: none, inaccessiblemem: none) uwtable
define dso_local i32 @foo() local_unnamed_addr #0 {
  %1 = load i32, ptr @x, align 4, !tbaa !3
  %2 = icmp eq i32 %1, 4
  %spec.store.select = select i1 %2, i32 4, i32 7
  store i32 %spec.store.select, ptr @x, align 4
  ret i32 %spec.store.select
}
```

Both these functions do not behave the same, in the unoptimized version, the function would always return 7, (which is what x would hold by the end of the function,) in the optimized version however the return value would depend upon the value of x at the time of function entry.

Reply:
Yes, this is expected, but only possible because the conditional store to `@x` can be speculated in this case.

Try replacing the clobbering store with something like `call void @clobber()` that may or may not overwrite the global in some unanalyzable way.

Reply:
Thanks a lot @nikic! Resolved this in the latest commit.